types/time.go

```diff
@@ -1563,234 +1607,67 @@ func extractSingleTimeValue(unit string, format string) (int64, int64, int64, fl
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	hours, err := strconv.ParseFloat(fields[0], 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	minutes, err := strconv.ParseFloat(fields[1], 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	seconds, err := strconv.ParseFloat(fields[2], 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	return 0, 0, 0, hours*float64(gotime.Hour) + minutes*float64(gotime.Minute) + seconds*float64(gotime.Second), nil
-}
-
-// extractHourMinute extracts hour and minute from a string and its format is `HH:MM`.
-func extractHourMinute(format string) (int64, int64, int64, float64, error) {
-	fields := strings.Split(format, ":")
-	if len(fields) != 2 {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	hours, err := strconv.ParseFloat(fields[0], 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	minutes, err := strconv.ParseFloat(fields[1], 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	return 0, 0, 0, hours*float64(gotime.Hour) + minutes*float64(gotime.Minute), nil
-}
-
-// extractDayMicrosecond extracts day and microsecond from a string and its format is `DD HH:MM:SS.FFFFFF`.
-func extractDayMicrosecond(format string) (int64, int64, int64, float64, error) {
-	fields := strings.Split(format, " ")
-	if len(fields) != 2 {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
-	}
-
-	days, err := strconv.ParseInt(fields[0], 10, 64)
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
+// extractTimeValue extracts years, months, days, microseconds from a string
+// MySQL permits any punctuation delimiter in the expr format.
+// See https://dev.mysql.com/doc/refman/8.0/en/expressions.html#temporal-intervals
+func extractTimeValue(format string, index, cnt int) (int64, int64, int64, float64, error) {
+	neg := false
+	originalFmt := format
+	format = strings.TrimSpace(format)
+	if len(format) > 0 && format[0] == '-' {
+		neg = true
+		format = format[1:]
 	}
-
-	_, _, _, value, err := extractHourMicrosecond(fields[1])
-	if err != nil {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
+	fields := make([]string, TimeValueCnt)
+	for i := range fields {
+		fields[i] = "0"
 	}
-
-	return 0, 0, days, value, nil
-}
-
-// extractDaySecond extracts day and hour from a string and its format is `DD HH:MM:SS`.
-func extractDaySecond(format string) (int64, int64, int64, float64, error) {
-	fields := strings.Split(format, " ")
-	if len(fields) != 2 {
-		return 0, 0, 0, 0, ErrIncorrectDatetimeValue.GenWithStackByArgs(format)
+	matches := numericRegex.FindAllString(format, -1)
+	if len(matches) > cnt {
```

Comment:
should it be `len(matches) != cnt` ?

Reply:
MySQL allows missing left parts of interval value. For example, if you specify a unit of DAY_SECOND, the value of expr is expected to have days, hours, minutes, and seconds parts. If you specify a value like '1:10', MySQL assumes that the days and hours parts are missing and the value represents minutes and seconds.

All of he following input are valid:

select date_add("2019-01-01 00:00:00",INTERVAL "" DAY_SECOND);
select date_add("2019-01-01 00:00:00",INTERVAL "1" DAY_SECOND);
select date_add("2019-01-01 00:00:00",INTERVAL "1 1" DAY_SECOND);
select date_add("2019-01-01 00:00:00",INTERVAL "1 1 1" DAY_SECOND);
select date_add("2019-01-01 00:00:00",INTERVAL "1 1 1 1" DAY_SECOND);

Take a look at this document
[Temporal Intervals](https://dev.mysql.com/doc/refman/8.0/en/expressions.html#temporal-intervals)

Reply:
okay